File: backend/src/entities/Email.ts
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  UpdateDateColumn,
  ManyToOne,
  ManyToMany,
  JoinTable
} from 'typeorm';
import { User } from './User';
import { Folder } from './Folder';
import { EmailAnalytics, EmailMetadata } from '../types/email';

export interface Recipient {
  name?: string;
  email: string;
}

export interface Attachment {
  filename: string;
  path: string;
  contentType: string;
  size: number;
}

@Entity('emails')
export class Email {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column()
  from: string;

  @Column('simple-array')
  to: string[];

  @Column({ nullable: true })
  subject: string;

  @Column('text')
  content: string;

  @Column('simple-array', { nullable: true })
  cc?: string[];

  @Column('simple-array', { nullable: true })
  bcc?: string[];

  @Column('jsonb', { nullable: true })
  attachments?: {
    filename: string;
    content: string;
    contentType: string;
    size: number;
  }[];

  @Column('jsonb', { nullable: true })
  metadata?: EmailMetadata;

  @Column('jsonb', { nullable: true })
  analytics?: EmailAnalytics;

  @Column('simple-array', { nullable: true })
  labels?: string[];

  @Column({ default: false })
  isRead: boolean;

  @Column({ default: false })
  isStarred: boolean;

  @Column({ default: false })
  isDraft: boolean;

  @Column({ default: false })
  isDeleted: boolean;

  @ManyToOne(() => User)
  user: User;

  @Column()
  userId: string;

  @ManyToMany(() => Folder, folder => folder.emails)
  folders: Folder[];

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;
} 

File: backend/src/models/email.py
from datetime import datetime
from typing import List, Optional, TYPE_CHECKING, ClassVar
from pydantic import BaseModel, EmailStr, Field, ConfigDict
from sqlalchemy import ForeignKey, String, Text, Boolean, DateTime, func, Integer
from sqlalchemy.orm import Mapped, mapped_column, relationship
import uuid

from src.db.base import Base
from .label import email_labels

if TYPE_CHECKING:
    from .user import User
    from .folder import Folder
    from .thread import Thread
    from .attachment import Attachment
    from .label import Label


class EmailBase(BaseModel):
    """Base email model."""
    subject: str = Field(..., min_length=1, max_length=200)
    content: str = Field(..., min_length=1)
    to_email: List[EmailStr]
    cc_email: Optional[List[EmailStr]] = []
    bcc_email: Optional[List[EmailStr]] = []
    attachments: Optional[List[str]] = []


class EmailCreate(EmailBase):
    """Email creation model."""
    pass


class EmailUpdate(BaseModel):
    """Email update model."""
    subject: Optional[str] = Field(None, min_length=1, max_length=200)
    content: Optional[str] = Field(None, min_length=1)
    to_email: Optional[List[EmailStr]] = None
    cc_email: Optional[List[EmailStr]] = None
    bcc_email: Optional[List[EmailStr]] = None
    attachments: Optional[List[str]] = None


class EmailInDB(EmailBase):
    """Email database model."""
    id: int
    user_id: int
    status: str = Field(..., pattern="^(draft|sent|failed)$")
    created_at: datetime
    updated_at: datetime
    sent_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class Email(Base):
    """Email model."""
    __tablename__ = "emails"

    # Admin configuration
    admin_list_display: ClassVar[list[str]] = [
        "id", "user_id", "thread_id", "subject", "from_address", "to_address", 
        "is_read", "is_sent", "is_draft", "is_trash", "sent_at"
    ]
    admin_searchable_fields: ClassVar[list[str]] = [
        "subject", "from_address", "to_address", "cc", "bcc"
    ]
    admin_filterable_fields: ClassVar[list[str]] = [
        "is_read", "is_sent", "is_draft", "is_trash"
    ]
    
    # Fields
    user_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    thread_id: Mapped[Optional[uuid.UUID]] = mapped_column(ForeignKey("threads.id", ondelete="CASCADE"), nullable=True, index=True)
    
    # Email content
    subject: Mapped[str] = mapped_column(String(255))
    body: Mapped[str] = mapped_column(Text)
    from_address: Mapped[str] = mapped_column(String(255))
    to_address: Mapped[str] = mapped_column(String(255))
    cc: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    bcc: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    
    # Email status
    is_read: Mapped[bool] = mapped_column(Boolean, default=False)
    is_sent: Mapped[bool] = mapped_column(Boolean, default=False)
    is_draft: Mapped[bool] = mapped_column(Boolean, default=True)
    is_trash: Mapped[bool] = mapped_column(Boolean, default=False)
    sent_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True), nullable=True)
    
    # Email importance
    importance: Mapped[int] = mapped_column(Integer, default=0)  # 0=normal, 1=important, 2=urgent
    
    # Relationships
    user: Mapped["User"] = relationship(back_populates="emails")
    thread: Mapped[Optional["Thread"]] = relationship(back_populates="emails")
    attachments: Mapped[List["Attachment"]] = relationship(back_populates="email", cascade="all, delete-orphan")
    labels: Mapped[List["Label"]] = relationship(
        secondary=email_labels,
        back_populates="emails",
        cascade="all, delete",
    )

    def __repr__(self) -> str:
        """String representation."""
        return f"<Email {self.id}: {self.subject}>"


class EmailResponse(BaseModel):
    """Email sending response model."""
    message_id: str
    status: str
    sent_at: datetime 